proxy/src/main/java/com/velocitypowered/proxy/plugin/VelocityPluginManager.java

```diff
@@ -105,8 +116,83 @@ public void loadPlugins(Path directory) throws IOException {
+          }
+        }
+      }
+
+      // match the to update plugin's id against already
+      // loaded plugins and replace them if match found
+      for (PluginDescription updatedDescription : updatesToApply) {
+        PluginDescription possibleMatch = found.get(updatedDescription.getId());
+        if (updatedDescription.getSource().isEmpty()) { //should not happen but just in case
+          logger.warn("No source found for plugin {} found", updatedDescription.getId());
+          continue;
+        }
+        Path oldPluginPath = null;
+        if (possibleMatch != null) {
+          if (possibleMatch.getSource().isEmpty()) {
+            logger.warn("No source for plugin {} found, continuing without update.",
+                    possibleMatch.getId());
+            continue;
+          }
+          //move old plugin to outdated plugin directory to rollback in case of failure
+          try {
+            oldPluginPath = possibleMatch.getSource().get();
+            Files.move(oldPluginPath, outdatedPluginDirectory.resolve(oldPluginPath.getFileName()),
+                    StandardCopyOption.REPLACE_EXISTING);
+          } catch (IOException e) {
+            logger.error("Unable to move plugin {} to outdated plugins folder at {}",
+                    possibleMatch.getId(), outdatedPluginDirectory.toString(), e);
+            continue;
+          }
+          logger.info("Moved plugin {} to outdated plugins folder at {}",
+                  possibleMatch.getId(), outdatedPluginDirectory.toString());
+        }
+        Path newPath = pluginDirectory.resolve(updatedDescription.getSource().get().getFileName());
+        try {
+          Files.move(updatedDescription.getSource().get(), newPath);
+          logger.info("Successfully updated plugin {} to version {}",
+                  updatedDescription.getId(), updatedDescription.getVersion());
+          PluginDescription movedDescription = loader.loadCandidate(newPath);
+          found.put(movedDescription.getId(), movedDescription);
+        } catch (Exception e) {
+          logger.error("Unable to update plugin {}", updatedDescription.getId(), e);
+          //rollback to old version if the plugin was a replacement and not newly added
+          if (oldPluginPath != null) {
+            Files.delete(updatedDescription.getSource().get());
+            Files.move(outdatedPluginDirectory.resolve(oldPluginPath.getFileName()), oldPluginPath);
+            logger.info("Rolled back plugin {} to version {}",
+                    updatedDescription.getId(), updatedDescription.getVersion());
+            try {
+              PluginDescription rolledBackDescription = loader.loadCandidate(oldPluginPath);
+              found.put(rolledBackDescription.getId(), rolledBackDescription);
+            } catch (Exception ex) {
+              logger.error("Unable to load rollback plugin candidate {}", oldPluginPath, ex);
+            }
+          }
+        }
+      }
+    }
```

Comment:
I think this block of code could be moved into a separate private method, to avoid overloading this method even more

Reply:
Should be resolved